### DISCUSSION TITLE
confusion about survey text

### DISCUSSION THREAD

**Discussion by cainesap** [jspsych/jsPsych#2322] [Q&A] [answered]
Hello,
Apologies because I think I've got myself confused about survey text and accessing question responses. Looking at other discussions on this site, I've pieced together the following:

```
// the survey
var askSubjectID = {
      type: 'survey-text',
      questions: [
          {prompt: 'Please enter the subject ID you were given'},
      ],
      data: {question_id: 'subjid'}
  }

// push this to the timeline...

// then at the end i'd like to put the subject ID in the filename
on_finish: function(){
        var sid = jsPsych.data.get().filter({question_id: 'subjid'}).values()[0];
        console.log(sid);
        var fileout = "jspsych_webgazer_data_subject_"+sid;
        saveData(fileout, jsPsych.data.get().csv());  // this is the save data function with 'write_data.php' demo'd on the jsPsysch site
    }
```
In the 'on_finish' function, 'sid' looks to me like JSON, so this may be an issue with me not knowing well enough how to parse JSON within JavaScript. In the console, the 'sid' object prints like this (I've tried to represent the indentation):

```
condition: "conditionA"
internal_node_id: "0.0-2.0"
question_id: "subjid"
response:
  Q0: "eggs"
  [[Prototype]]: Object
rt: 1556.2999999970198
subject: undefined
time_elapsed: 4687
trial_index: 2
trial_type: "survey-text"
[[Prototype]]: Object
```

I've tried to get at Q0 with 'sid.response.Q0' and variants around that, but so far haven't succeeded. Any pointers here gratefully received!

I'm using 6.3.0 btw; maybe I should upgrade to 7 (but I'm also using js-psychophysics which is compatible with 6.3.0 and I'm not sure it's been updated for 7).

thanks as ever for jsPsych and thanks in advance for any help! Andrew

**Comment by becky-gilbert** [comment, marked as answer]
Hi @cainesap! You're very close. This line:
```js
var sid = jsPsych.data.get().filter({question_id: 'subjid'}).values()[0];
```
Is getting all of the jsPsych data, then filtering the trials based on the value of "question_id", then returning those trials as an array, and finally giving you the data object from the first (probably only?) trial in that array. So in other words it's giving you all of the data from that trial. Now what you need to do is access the "response" property from that trial data, and then access the specific question name ("Q0") within "response". So try this:
```js
var sid = jsPsych.data.get().filter({question_id: 'subjid'}).values()[0].response.Q0;
```
Does that work for you?

Btw you're right that certain types of data used to be stored as JSON strings in the jsPsych data, but then we switched to storing everything as objects/arrays rather than JSON. So with V6.3.0 I don't think you need the JSON parsing step. In earlier jsPsych versions you would need to parse the "response" with `JSON.parse()` before getting the value of Q0.

**Reply by cainesap** [reply]
Hi @becky-gilbert thank you so much for your reply! That is great, it does work as I'd been looking for :)
In case someone else finds this question and it's useful, I generalised it for use in `on_trial_finish` with a try..catch statement to handle errors until the question has been asked and answered (tell me if this looks horrible for any reason!) :

```
on_trial_finish: function(){
    var sid = jsPsych.data.getLastTrialData().filter({question_id: 'subjid'}).values()[0];
    try {
        console.log(sid.response.Q0);
        var fileout = "jspsych_webgazer_data_subject_"+sid.response.Q0;
        saveData(fileout, jsPsych.data.get().csv());  // this is the save data function with 'write_data.php' demo'd on the jsPsysch site
    } catch (error) {
        console.log("No subject ID as yet");
    }
}
```

thanks again, Andrew

**Reply by becky-gilbert** [reply]
Thanks for the update and for sharing your solution with others 😀
